### PR TITLE
online dpo trainer based on rloo trainer

### PR DESCRIPTION
**Pull request by mnoukhov**

### CODE REVIEW

File: trl/trainer/online_dpo_trainer.py [hunk 1]
```diff
@@ -0,0 +1,619 @@
...
+                    for i in range(0, queries.shape[0], args.local_rollout_forward_batch_size):
+                        query = queries[i : i + args.local_rollout_forward_batch_size]
+                        query_response, logits = generate(
+                            unwrapped_model,
+                            query,
+                            tokenizer.pad_token_id,
+                            generation_config,
+                        )
+                        response = query_response[:, context_length:]
+                        logits /= args.temperature + 1e-7
```
Comment (vwxyzjn) [inline trl/trainer/online_dpo_trainer.py:301]: You don't need to adjust the logits by temperature here because it's already adjusted in the `generate` function. 

```suggestion
```

File: trl/trainer/online_dpo_trainer.py [hunk 2]
```diff
@@ -0,0 +1,619 @@
...
+                    saved_data["prompt"].extend(gather_object(decoded_queries))
+                    saved_data["chosen"].extend(gather_object(decoded_chosen))
+                    saved_data["rejected"].extend(gather_object(decoded_rejected))
+                    saved_data["update"].extend(gather_object([update for _ in range(num_examples)]))
+
+                torch.cuda.empty_cache()
+
+            # Do multiple epochs of PPO training, with a fresh random shuffle in each epoch
+            for ppo_epoch_idx in range(args.num_ppo_epochs):
+                b_inds = np.arange(args.local_batch_size)
```
Comment (vwxyzjn) [inline trl/trainer/online_dpo_trainer.py:399]:
```suggestion
                b_inds = np.random.permutation(args.local_batch_size)
```

**Review by vwxyzjn — COMMENTED** [review]
Very nice work and thanks for the PR @mnoukhov! Some quick comments.

File: trl/trainer/online_dpo_trainer.py (hunk 1 above)
Comment (mnoukhov) [reply trl/trainer/online_dpo_trainer.py]: Does this mean I don't need to adjust it in other cases too? Or just here?

**Comment by vwxyzjn** [comment]
Closing in favor of #1809
